### PR TITLE
add the portability sample and allow for enabling globally during instance creation

### LINKED ISSUE #343: Add support for VK_KHR_portability_subset extension
The `VK_KHR_portability_subset` extension, being developed by the [Vulkan Portability initiative](https://github.com/KhronosGroup/Vulkan-Portability), is currently provisional, but is approaching finalization, and is expected to be finalized and released in early 2022.

In conjunction with that release, it would be helpful for Vulkan-Samples to support that finalized extension at several distinct levels:

1. Provide framework support to look for the presence of the [`VK_KHR_portability_subset`](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VK_KHR_portability_subset.html) extension in a Vulkan implementation, query for the `pNext` structures provided by that extension, and store the content of those structures within the framework, so portability features and limits will be available to all samples.

2. Provide at least one new sample, targeted specifically to demonstrate how to use the `VK_KHR_portability_subset` extension to query for a portability platform feature, and provide two application code paths to use that feature if it exists, or work around the absence of the feature if it doesn't exist.

3. Modify existing samples that may already be using platform `#ifdef`-style code checks to follow separate code paths for portability platforms (eg- Apple platforms), to accommodate existing platform limitations, and replace those `#ifdef`-style code checks with queries of the `VK_KHR_portability_subset` extension feature flags to choose between sample code paths.

### PR DESCRIPTION
**Pull request by gpx1000**
## Description

Add portability sample.  NB: this requires Vulkan 1.3.

Fixes #343 

## General Checklist:

Please ensure the following points are checked:

- [x ] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [ x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [ x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [ x] I have commented any added functions (in line with Doxygen)
- [x ] I have commented any code that could be hard to understand
- [x ] My changes do not add any new compiler warnings
- [ x] My changes do not add any new validation layer errors or warnings
- [ x] I have used existing framework/helper functions where possible
- [ x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x ] My changes do not add any regressions
- [ x] I have tested every sample to ensure everything runs correctly
- [ x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [ x] I have tested the sample on at least one compliant Vulkan implementation
- [ ] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x ] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [ x] Any dependent assets have been merged and published in downstream modules
- [ ] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

### CODE REVIEW

**Review by gary-sweet — COMMENTED** [review]
Correctly reports "not supported" for me, but that's all I can test.

File: samples/extensions/portability/README.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+This tutorial, along with the accompanying example code, demonstrates the use of the [VK_KHR_portability_subset](https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VK_KHR_portability_subset) extension. 
+When the VK_INSTANCE_CREATE_ENUMERATE_PORTABILITY_BIT_KHR is set in the ```Instance``` class, Vulkan will consider 
+devices that aren't fully conformant such as [MoltenVk](https://github.com/KhronosGroup/MoltenVK) to be identified 
+as a conformant implementation.  When this happens, use the VkPhysicalDevicePortabilitySubsetPropertiesKHR extension 
+with the [vkGetPhysicalDeviceFeatures2](https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#vkGetPhysicalDeviceFeatures2)
+as detailed below to get the list of supported/unsupported features.
+
+This tutorial along with the accompanying code also demonstrates the use of the [VkPhysicalDevicePortabilitySubsetPropertiesKHR]
+(https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.
```
Comment (SaschaWillems) [inline samples/extensions/portability/README.md]: It seems the space/line break here after the `vkspec.` breaks the link in markdown.

Comment (gpx1000) [reply samples/extensions/portability/README.md]: thanks for the catch.  Should be fixed.

**Review by TomAtkinsonArm — COMMENTED** [review]
LGTM - Shows expected behavior on my desktop

Will need to test on a MoltenVK Device at some point

**Comment by billhollings** [comment]
@gpx1000 

I'm trying to build and test this. Portability is not getting enabled.

The `portability/README.md` says that both `VK_ENABLE_BETA_EXTENSIONS` and `VKB_ENABLE_PORTABILITY` define macros need to be enabled during build. 

Unfortunately, I'm not a `cmake` guru. How exactly should those macros be defined within the documented Vulkan-Samples command line build instructions listed here?

```
cmake -H. -Bbuild/mac -DCMAKE_BUILD_TYPE=Release
cmake --build build/mac --config Release --target vulkan_samples -- -j4
```

Thanks.

**Comment by gpx1000** [comment]
cmake -DCMAKE_BUILD_TYPE=Debug  -DVKB_VALIDATION_LAYERS=ON -DVKB_ENABLE_PORTABILITY=ON -DVK_ENABLE_BETA_EXTENSIONS=ON ....

modify as your needs exhibit :)  i.e. you likely want to tell it something like a XCode generation is required or what you're targeting like ios toolchain what compiler etc.

**Comment by billhollings** [comment]
>`cmake -DCMAKE_BUILD_TYPE=Debug -DVKB_VALIDATION_LAYERS=ON -DVKB_ENABLE_PORTABILITY=ON -DVK_ENABLE_BETA_EXTENSIONS=ON ....`

Hmmm...I tried that...specifically...

```
cmake -H. -Bbuild/mac -DCMAKE_BUILD_TYPE=Release -DVK_ENABLE_BETA_EXTENSIONS=ON -DVKB_ENABLE_PORTABILITY=ON
cmake --build build/mac --config Release --target vulkan_samples -- -j4
```

but got the following results during the first command...

```
CMake Warning:
  Manually-specified variables were not used by the project:

    VKB_ENABLE_PORTABILITY
    VK_ENABLE_BETA_EXTENSIONS
```

and the portability sample continues to report onscreen `VK_ENABLE_BETA_EXTENSIONS not enabled can't list portability feature set`

**Comment by gpx1000** [comment]
cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_DEPENDS_USE_COMPILER=FALSE -G "CodeBlocks - Unix Makefiles" /Users/swinston/Projects/holochip/vulkan_samples --target vulkan_samples -- -j4 -DVK_ENABLE_BETA_EXTENSIONS=ON -DVKB_ENABLE_PORTABILITY=ON

Another way to do it is above.  The -- at the end of the line tells CMake to pass remainder directly to the compiler.  That's the easy way.  The explicit way is to define the CMAKE_C_FLAGS and CMAKE_CXX_FLAGS to hold the values you want them to i.e. like this:

cmake -H. -Bbuild/mac -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_FLAGS="-DVK_ENABLE_BETA_EXTENSIONS=ON -DVKB_ENABLE_PORTABILITY=ON" -DCMAKE_CXX_FLAGS="-DVK_ENABLE_BETA_EXTENSIONS=ON -DVKB_ENABLE_PORTABILITY=ON"
cmake --build build/mac --config Release --target vulkan_samples -- -j4

NB: the above will use the system defined code generator.  My original command up top is using CodeBlocks -Unix Makefiles for the generator.  There's a possibility that's not what you want and instead want to use XCode or target iOS.  If that applies to you, you'd want to modify the -G target correspondingly.  Also, you might need to supply a toolchain file depending upon how you have your system setup (-T then toolchain file).  So, if I wanted to target iOS, I might use -G XCode -T ios.toolchain.  This also describes how you'd target Android if you want but that's beyond the scope of this quick intro.

**Comment by billhollings** [comment]
Thanks. Finally got this working with your second build option (`CMAKE_C_FLAGS`, etc).

Confirmed this sample works correctly with MoltenVK. The feature flags reported by the sample match those returned by MoltenVK.

**Review by billhollings — APPROVED** [review]

**Comment by marty-johnson59** [comment]
@billhollings, if you're happy with this, I'm tempted to merge.  Anyone else you'd like to see review this first?

**Comment by SaschaWillems** [comment]
I'm fine with merging. I don't have an implementation to test this on, but code compiles, reports extension as not supported and both bode and documentation look fine :)

**Comment by marty-johnson59** [comment]
OK, I'll merge - this is a bit of a special case.  LMK if issues and we can revert if needed.

**Pull request merged**